webpack/assets/javascripts/react_app/components/common/IndexPageTemplate/index.js

```diff
@@ -0,0 +1,206 @@
+    total_entries = 0,
+    global_actions,
+    primary_action,
+    before_toolbar_content,
+    empty_state,
+    title,
+  } = useSelector(state => selectAPIResponse(state, key));
+  const error = useSelector(state => selectAPIErrorMessage(state, key));
+  const status = useSelector(state => selectAPIStatus(state, key));
+
+  const dispatch = useDispatch();
+  const reloadData = () =>
+    dispatch(fetchData(path, { page, perPage, query, sortBy, sortOrder }));
+
+  useEffect(() => {
+    reloadData();
+  }, []);
+
+  useEffect(() => {
+    if (status === STATUS.RESOLVED && rows && columns) {
+      const isActions = rows.some(({ actions }) => !!actions);
+      setTableRows(handleRows(rows));
+      setTableColumns(handleColumns(columns, isActions));
+    }
+  }, [rows, columns, status]);
+
+  const handleRows = serverRows =>
+    serverRows.asMutable({ deep: true }).map(({ cells, actions }) => {
+      const nextCells = cells.map(cell => {
+        if (typeof cell !== 'object') return cell;
+        if (cell.type === 'link') {
+          return (
+            <TableText>
+              <a href={cell.path}>{cell.label}</a>
+            </TableText>
+          );
+        }
+        return cell;
+      });
+      actions &&
+        nextCells.push({
+          title: <ActionsDropdown actions={actions} reloadData={reloadData} />,
+        });
+
+      return { cells: nextCells };
+    });
+
+  const handleColumns = (serverColumns, isActions) => {
+    const nextColumns = serverColumns.asMutable({ deep: true }).map(colData => {
+      if (typeof colData !== 'object') {
+        return { title: colData };
+      }
+      const { sort_by, width, label } = colData;
+      const col = { title: label, transforms: [] };
+      if (sort_by) {
+        col.sortKey = sort_by;
+        col.transforms = [sortable];
+      }
+      if (width) {
+        col.transforms = [...col.transforms, cellWidth(width)];
+      }
+      return col;
+    });
+    isActions && nextColumns.push(__('Actions'));
+    return nextColumns;
+  };
+
+  const beforeToolbarComponent = before_toolbar_content && (
+    <Text
+      className="index-page-description"
+      dangerouslySetInnerHTML={{ __html: before_toolbar_content }}
+    />
+  );
+
+  if (total_entries === 0 && empty_state) {
+    return (
+      <EmptyState
+        {...empty_state}
+        description={
+          <div
+            dangerouslySetInnerHTML={{
+              __html: empty_state.description,
+            }}
+          />
+        }
+      />
+    );
+  }
+
+  const paginationProps = variant => ({
+    id: `index-page-${variant}-pagination`,
+    itemCount: total_entries,
+    perPage,
+    page,
+    variant: PaginationVariant[variant],
+    onSetPage: (e, pageNo) => dispatch(onTableSetPage(pageNo, path)),
+    onPerPageSelect: (e, perP) => dispatch(onTablePerPageSelect(perP, path)),
+    perPageOptions: getPerPageOptions(perPage, defaultPerPage),
+  });
+
+  return (
+    <PageLayout
```

Comment:
This is a PF3 based component, so this template will be mixed between pf3 and pf4 (and due to the fact this is a template, we will increase our dependency over PF3 and makes our technical dept deepr)

On the host details page, I consumed the <PageSection> component from PF4, how complicated it would be to create a  PF4 `PageSection` wrapper for this template?

Reply:
Thanks, @amirfefer you are right,
I will address that soon in https://projects.theforeman.org/issues/32991
in a parallel PR, as it may affect many existing pages